core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -364,30 +368,98 @@ class ReplicaManager(val config: KafkaConfig,
-          s"${stopReplicaRequest.controllerEpoch}. Latest known controller epoch is $controllerEpoch")
+        stateChangeLogger.warn(s"Ignoring StopReplica request (delete=$deletePartition) from " +
+          s"controller $controllerId with correlation id $correlationId " +
+          s"since its controller epoch ${stopReplicaRequest.controllerEpoch} is old. " +
+          s"Latest known controller epoch is $controllerEpoch")
         (responseMap, Errors.STALE_CONTROLLER_EPOCH)
       } else {
-        val partitions = stopReplicaRequest.partitions.asScala.toSet
+        val partitions = mutable.Set.empty[TopicPartition]
         controllerEpoch = stopReplicaRequest.controllerEpoch
+
+        stopReplicaRequest.topicStates.asScala.foreach { topicState =>
+          topicState.partitionStates.asScala.foreach { partitionState =>
+            val topicPartition = new TopicPartition(topicState.topicName, partitionState.partitionIndex)
+
+            getPartition(topicPartition) match {
+              case HostedPartition.Offline =>
+                stateChangeLogger.warn(s"Ignoring StopReplica request (delete=$deletePartition) from " +
+                  s"controller $controllerId with correlation id $correlationId " +
+                  s"epoch $controllerEpoch for partition $topicPartition as the local replica for the " +
+                  "partition is in an offline log directory")
+                responseMap.put(topicPartition, Errors.KAFKA_STORAGE_ERROR)
+
+              case HostedPartition.Online(partition) =>
+                val currentLeaderEpoch = partition.getLeaderEpoch
+                val requestLeaderEpoch = partitionState.leaderEpoch
+                // When a topic is deleted, the leader epoch is not incremented. To circumvent this,
+                // a sentinel value (EpochDuringDelete) overwriting any previous epoch is used.
+                // When an older version of the StopReplica request which does not contain the leader
+                // epoch, a sentinel value (NoEpoch) is used and bypass the epoch validation.
+                if (requestLeaderEpoch == LeaderAndIsr.EpochDuringDelete ||
+                    requestLeaderEpoch == LeaderAndIsr.NoEpoch ||
+                    requestLeaderEpoch > currentLeaderEpoch) {
+                  partitions += topicPartition
+                } else if (requestLeaderEpoch < currentLeaderEpoch) {
+                  stateChangeLogger.warn(s"Ignoring StopReplica request (delete=$deletePartition) from " +
+                    s"controller $controllerId with correlation id $correlationId " +
+                    s"epoch $controllerEpoch for partition $topicPartition since its associated " +
+                    s"leader epoch $requestLeaderEpoch is smaller than the current " +
+                    s"leader epoch $currentLeaderEpoch")
+                  responseMap.put(topicPartition, Errors.STALE_CONTROLLER_EPOCH)
```

Comment:
Hmm.. This error seems a little inaccurate. Could we use `FENCED_LEADER_EPOCH`?

Reply:
Good point. I completely forgot to raise it...

I use `STALE_CONTROLLER_EPOCH` here to stay inline with the `LeaderAndIsr` API which uses is as well when the leader epoch is stale. See here: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ReplicaManager.scala#L1227

It was introduced in an old refactoring: https://github.com/apache/kafka/commit/a9ff3f2eced76a0aa13804439101c2897916e250#diff-4f99f5a41c14e2a8523c03ce4ae23987L630. It seems that back in the days, we had `StaleLeaderEpochCode` but it got replaced by `STALE_CONTROLLER_EPOCH `.

I was actually wondering if we should stay inline with the current behavior of the `LeaderAndIsr` or just use `FENCED_LEADER_EPOCH`.  If there a not other reasons besides the historical one to use `STALE_CONTROLLER_EPOCH`, `FENCED_LEADER_EPOCH` sounds indeed more appropriate.

Reply:
I have changed it to use `FENCED_LEADER_EPOCH`.